<h3 id="BedToIntervalList">BedToIntervalList</h3>
<section>
<p>Converts a BED file to a Picard Interval List.  This tool provides easy conversion from BED to the Picard interval_list format which is required by many Picard processing tools. Note that the coordinate system of BED files is such that the first base or position in a sequence is numbered "0", while in interval_list files it is numbered "1".<br /><br />BED files contain sequence data displayed in a flexible format that includes nine optional fields, in addition to three required fields within the annotation tracks. The required fields of a BED file include:<pre>     chrom - The name of the chromosome (e.g. chr20) or scaffold (e.g. scaffold10671) <br />     chromStart - The starting position of the feature in the chromosome or scaffold. The first base in a chromosome is numbered "0" <br />     chromEnd - The ending position of the feature in the chromosome or scaffold.  The chromEnd base is not included in the display of the feature. For example, the first 100 bases of a chromosome are defined as chromStart=0, chromEnd=100, and span the bases numbered 0-99.</pre>In each annotation track, the number of fields per line must be consistent throughout a data set. For additional information regarding BED files and the annotation field options, please see: http://genome.ucsc.edu/FAQ/FAQformat.html#format1.<br /> <br /> Interval_list files contain sequence data distributed into intervals. The interval_list file format is relatively simple and reflects the SAM alignment format to a degree.  A SAM style header must be present in the file that lists the sequence records against which the intervals are described.  After the header, the file then contains records, one per line in plain text format with the following values tab-separated::<pre>      -Sequence name (SN) - The name of the sequence in the file for identification purposes, can be chromosome number e.g. chr20 <br />      -Start position - Interval start position (starts at +1) <br />      -End position - Interval end position (1-based, end inclusive) <br />      -Strand - Indicates +/- strand for the interval (either + or -) <br />      -Interval name - (Each interval should have a unique name) </pre><br/>This tool requires a sequence dictionary, provided with the SEQUENCE_DICTIONARY or SD argument. The value given to this argument can be any of the following:<pre>    - A file with .dict extension generated using Picard's CreateSequenceDictionaryTool</br>    - A reference.fa or reference.fasta file with a reference.dict in the same directory</br>    - Another IntervalList with @SQ lines in the header from which to generate a dictionary</br>    - A VCF that contains #contig lines from which to generate a sequence dictionary</br>    - A SAM or BAM file with @SQ lines in the header from which to generate a dictionary</br></pre><h4>Usage example:</h4><pre>java -jar picard.jar BedToIntervalList \<br />      I=input.bed \<br />      O=list.interval_list \<br />      SD=reference_sequence.dict</pre><br /> <br /> <hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>INPUT (File)</td><td>The input BED file  Required. </td></tr>
<tr><td>SEQUENCE_DICTIONARY (File)</td><td>The sequence dictionary, or BAM/VCF/IntervalList from which a dictionary can be extracted.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>The output Picard Interval List  Required. </td></tr>
<tr><td>SORT (Boolean)</td><td>If true, sort the output interval list before writing it.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>UNIQUE (Boolean)</td><td>If true, unique the output interval list by merging overlapping regions, before writing it (implies sort=true).  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
</table>
</section>
