<a id="CollectRnaSeqMetrics"/>
<h3>CollectRnaSeqMetrics</h3>
<section>
<p>Collect metrics about the alignment of RNA to various functional classes of loci in the genome:coding, intronic, UTR, intergenic, ribosomal. Also determines strand-specificity for strand-specific libraries.</p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>REF_FLAT (File)</td><td>Gene annotations in refFlat form.  Format described here: http://genome.ucsc.edu/goldenPath/gbdDescriptionsOld.html#RefFlat  Required. </td></tr>
<tr><td>RIBOSOMAL_INTERVALS (File)</td><td>Location of rRNA sequences in genome, in interval_list format.  If not specified no bases will be identified as being ribosomal.  Format described here: http://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/samtools/util/IntervalList.html  Default value: null. </td></tr>
<tr><td>STRAND_SPECIFICITY (StrandSpecificity)</td><td>For strand-specific library prep. For unpaired reads, use FIRST_READ_TRANSCRIPTION_STRAND if the reads are expected to be on the transcription strand.  Required. Possible values: {NONE, FIRST_READ_TRANSCRIPTION_STRAND, SECOND_READ_TRANSCRIPTION_STRAND} </td></tr>
<tr><td>MINIMUM_LENGTH (Integer)</td><td>When calculating coverage based values (e.g. CV of coverage) only use transcripts of this length or greater.  Default value: 500. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>CHART_OUTPUT (File)</td><td>The PDF file to write out a plot of normalized position vs. coverage.  Default value: null. </td></tr>
<tr><td>IGNORE_SEQUENCE (String)</td><td>If a read maps to a sequence specified with this option, all the bases in the read are counted as ignored bases.  These reads are not counted as   Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>RRNA_FRAGMENT_PERCENTAGE (Double)</td><td>This percentage of the length of a fragment must overlap one of the ribosomal intervals for a read or read pair by this must in order to be considered rRNA.  Default value: 0.8. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>METRIC_ACCUMULATION_LEVEL (MetricAccumulationLevel)</td><td>The level(s) at which to accumulate metrics.    Default value: [ALL_READS]. This option can be set to 'null' to clear the default value. Possible values: {ALL_READS, SAMPLE, LIBRARY, READ_GROUP} This option may be specified 0 or more times. This option can be set to 'null' to clear the default list. </td></tr>
<tr><td>INPUT (File)</td><td>Input SAM or BAM file.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>File to write the output to.  Required. </td></tr>
<tr><td>ASSUME_SORTED (Boolean)</td><td>If true (default), then the sort order in the header file will be ignored.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>STOP_AFTER (Long)</td><td>Stop after processing N reads, mainly for debugging.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
