<a id="CollectAlignmentSummaryMetrics"/>
<h3>CollectAlignmentSummaryMetrics</h3>
<section>
<p>Produces from a SAM or BAM a file containing summary alignment metrics</p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>MAX_INSERT_SIZE (Integer)</td><td>Paired end reads above this insert size will be considered chimeric along with inter-chromosomal pairs.  Default value: 100000. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>ADAPTER_SEQUENCE (String)</td><td>List of adapter sequences to use when processing the alignment metrics  This option may be specified 0 or more times. This option can be set to 'null' to clear the default list. </td></tr>
<tr><td>METRIC_ACCUMULATION_LEVEL (MetricAccumulationLevel)</td><td>The level(s) at which to accumulate metrics.    Possible values: {ALL_READS, SAMPLE, LIBRARY, READ_GROUP} This option may be specified 0 or more times. This option can be set to 'null' to clear the default list. </td></tr>
<tr><td>IS_BISULFITE_SEQUENCED (Boolean)</td><td>Whether the SAM or BAM file consists of bisulfite sequenced reads.    Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INPUT (File)</td><td>Input SAM or BAM file.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>File to write the output to.  Required. </td></tr>
<tr><td>REFERENCE_SEQUENCE (File)</td><td>Reference sequence fasta  Default value: null. </td></tr>
<tr><td>ASSUME_SORTED (Boolean)</td><td>If true (default), then the sort order in the header file will be ignored.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>STOP_AFTER (Long)</td><td>Stop after processing N reads, mainly for debugging.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
