<h3 id="CollectBaseDistributionByCycle">CollectBaseDistributionByCycle</h3>
<section>
<p>Chart the nucleotide distribution per cycle in a SAM or BAM fileThis tool produces a chart of the nucleotide distribution per cycle in a SAM or BAM file in order to enable assessment of systematic errors at specific positions in the reads.<br /><br /><h4>Interpretation notes</h4>Increased numbers of miscalled bases will be reflected in base distribution changes and increases in the number of Ns. In general, we expect that for any given cycle, or position within reads, the relative proportions of A, T, C and G should reflect the AT:GC content of the organism's genome.  Thus, for all four nucleotides, flattish lines would be expected.  Deviations from this expectation, for example a spike of A at a particular cycle (position within reads), would suggest a systematic sequencing error.<h4>Note on quality trimming</h4>In the past, many sequencing data processing workflows included discarding the low-quality tails of reads by applying hard-clipping at some arbitrary base quality threshold value. This is no longer useful because most sophisticated analysis tools (such as the GATK variant discovery tools) are quality-aware, meaning that they are able to take base quality into account when weighing evidence provided by sequencing reads. Unnecessary clipping may interfere with other quality control evaluations and may lower the quality of analysis results. For example, trimming reduces the effectiveness of the Base Recalibration (BQSR) pre-processing step of the <a href='https://www.broadinstitute.org/gatk/guide/best-practices'>GATK Best Practices for Variant Discovery</a>, which aims to correct some types of systematic biases that affect the accuracy of base quality scores. <br /><h4>Usage example:</h4><pre>java -jar picard.jar CollectBaseDistributionByCycle \<br />      CHART=collect_base_dist_by_cycle.pdf \<br />      I=input.bam \<br />      O=output.txt</pre><hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>CHART_OUTPUT (File)</td><td>A file (with .pdf extension) to write the chart to.  Required. </td></tr>
<tr><td>ALIGNED_READS_ONLY (Boolean)</td><td>If set to true, calculate the base distribution over aligned reads only.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>PF_READS_ONLY (Boolean)</td><td>If set to true, calculate the base distribution over PF reads only (Illumina specific). PF reads are reads that passed the internal quality filters applied by Illumina sequencers.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INPUT (File)</td><td>Input SAM or BAM file.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>File to write the output to.  Required. </td></tr>
<tr><td>ASSUME_SORTED (Boolean)</td><td>If true (default), then the sort order in the header file will be ignored.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>STOP_AFTER (Long)</td><td>Stop after processing N reads, mainly for debugging.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
