<h3 id="CollectTargetedPcrMetrics">CollectTargetedPcrMetrics</h3>
<section>
<p>Calculate PCR-related metrics from targeted sequencing data. This tool calculates a set of PCR-related metrics from an aligned SAM or BAM file containing targeted sequencing data. It is appropriate for data produced with multiple small-target technologies including exome sequencing an custom amplicon panels such as the Illumina <a href='http://www.illumina.com/content/dam/illumina-marketing/documents/products/datasheets/datasheet_truseq_custom_amplicon.pdf'>TruSeq Custom Amplicon (TSCA)</a> kit. <br /><br />If a reference sequence is provided, AT/GC dropout metrics will be calculated and the PER_TARGET_COVERAGE  option can be used to output GC content and mean coverage information for each target. The AT/GC dropout metrics indicate the degree of inadequate coverage of a particular region based on its AT or GC content. The PER_TARGET_COVERAGE option can be used to output GC content and mean sequence depth information for every target interval. <br /><br />Please note that coverage depth at each locus should not exceed a limit of java MAX_SHORT ~32K.  This is because CollectTargetedPcrMetrics tool uses a short array to calculate coverage metrics.<h4>Usage example:</h4><pre>java -jar picard.jar CollectTargetedPcrMetrics \<br />      I=input.bam \<br />      O=pcr_metrics.txt \<br />      R=reference_sequence.fasta \<br />      AMPLICON_INTERVALS=amplicon.interval_list \<br />      TARGET_INTERVALS=targets.interval_list </pre>For explanations of the output metrics, see http://broadinstitute.github.io/picard/picard-metric-definitions.html#TargetedPcrMetrics<hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>AMPLICON_INTERVALS (File)</td><td>An interval list file that contains the locations of the baits used.  Required. </td></tr>
<tr><td>CUSTOM_AMPLICON_SET_NAME (String)</td><td>Custom amplicon set name. If not provided it is inferred from the filename of the AMPLICON_INTERVALS intervals.  Default value: null. </td></tr>
<tr><td>TARGET_INTERVALS (File)</td><td>An interval list file that contains the locations of the targets.  Default value: null. This option must be specified at least 1 times. </td></tr>
<tr><td>INPUT (File)</td><td>An aligned SAM or BAM file.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>The output file to write the metrics to.  Required. </td></tr>
<tr><td>METRIC_ACCUMULATION_LEVEL (MetricAccumulationLevel)</td><td>The level(s) at which to accumulate metrics.  Default value: [ALL_READS]. This option can be set to 'null' to clear the default value. Possible values: {ALL_READS, SAMPLE, LIBRARY, READ_GROUP} This option may be specified 0 or more times. This option can be set to 'null' to clear the default list. </td></tr>
<tr><td>PER_TARGET_COVERAGE (File)</td><td>An optional file to output per target coverage information to.  Default value: null. </td></tr>
<tr><td>PER_BASE_COVERAGE (File)</td><td>An optional file to output per base coverage information to. The per-base file contains one line per target base and can grow very large. It is not recommended for use with large target sets.  Default value: null. </td></tr>
<tr><td>NEAR_DISTANCE (Integer)</td><td>The maximum distance between a read and the nearest probe/bait/amplicon for the read to be considered 'near probe' and included in percent selected.  Default value: 250. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>MINIMUM_MAPPING_QUALITY (Integer)</td><td>Minimum mapping quality for a read to contribute coverage.  Default value: 1. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>MINIMUM_BASE_QUALITY (Integer)</td><td>Minimum base quality for a base to contribute coverage.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>CLIP_OVERLAPPING_READS (Boolean)</td><td>True if we are to clip overlapping reads, false otherwise.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>COVERAGE_CAP (Integer)</td><td>Parameter to set a max coverage limit for Theoretical Sensitivity calculations. Default is 200.  Default value: 200. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>SAMPLE_SIZE (Integer)</td><td>Sample Size used for Theoretical Het Sensitivity sampling. Default is 10000.  Default value: 10000. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
