<h3 id="CollectQualityYieldMetrics">CollectQualityYieldMetrics</h3>
<section>
<p>Collect metrics about reads that pass quality thresholds and Illumina-specific filters.  This tool evaluates the overall quality of reads within a bam file containing one read group. The output indicates the total numbers of bases within a read group that pass a minimum base quality score threshold and (in the case of Illumina data) pass Illumina quality filters as described in the <a href='https://www.broadinstitute.org/gatk/guide/article?id=6329'>GATK Dictionary entry</a>. <br /><h4>Note on base quality score options</h4>If the quality score of read bases has been modified in a previous data processing step such as <a href='https://www.broadinstitute.org/gatk/guide/article?id=44'>GATK Base Recalibration</a> and an OQ tag is available, this tool can be set to use the OQ value instead of the primary quality value for the evaluation. <br /><br />Note that the default behaviour of this program changed as of November 6th 2015 to no longer include secondary and supplemental alignments in the computation. <br /><h4>Usage Example:</h4><pre>java -jar picard.jar CollectQualityYieldMetrics \<br />       I=input.bam \<br />       O=quality_yield_metrics.txt \<br /></pre>Please see <a href='https://broadinstitute.github.io/picard/picard-metric-definitions.html#CollectQualityYieldMetrics.QualityYieldMetrics'>the QualityYieldMetrics documentation</a> for details and explanations of the output metrics.<hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>USE_ORIGINAL_QUALITIES (Boolean)</td><td>If available in the OQ tag, use the original quality scores as inputs instead of the quality scores in the QUAL field.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INCLUDE_SECONDARY_ALIGNMENTS (Boolean)</td><td>If true, include bases from secondary alignments in metrics. Setting to true may cause double-counting of bases if there are secondary alignments in the input file.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INCLUDE_SUPPLEMENTAL_ALIGNMENTS (Boolean)</td><td>If true, include bases from supplemental alignments in metrics. Setting to true may cause double-counting of bases if there are supplemental alignments in the input file.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INPUT (File)</td><td>Input SAM or BAM file.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>File to write the output to.  Required. </td></tr>
<tr><td>ASSUME_SORTED (Boolean)</td><td>If true (default), then the sort order in the header file will be ignored.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>STOP_AFTER (Long)</td><td>Stop after processing N reads, mainly for debugging.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
