<h3 id="CollectSequencingArtifactMetrics">CollectSequencingArtifactMetrics</h3>
<section>
<p>Collect metrics to quantify single-base sequencing artifacts.This tool examines two sources of sequencing errors resulting from hybrid selection protocols: <a href='https://www.broadinstitute.org/gatk/guide/article?id=6333'>bait-bias</a> and <a href='https://www.broadinstitute.org/gatk/guide/article?id=6332'>pre-adapter artifacts</a>. For a brief primer on these types of artifacts, see the corresponding GATK Dictionary entries.<br /><br />This tool produces four files; summary and detail metrics files for both pre-adapter and bait-bias artifacts. The detailed metrics show the error rates for each type of base substitution within every possible triplet base configuration.  Error rates associated with these substitutions are Phred-scaled and provided as quality scores, the lower the value, the more likely it is that an alternate base call is due to an artifact. The summary metrics provide likelihood information on the "worst-case" errors. <br /><h4>Usage example:</h4><pre>java -jar picard.jar CollectSequencingArtifactMetrics \<br />     I=input.bam \<br />     O=artifact_metrics.txt \<br />     R=reference_sequence.fasta</pre>For additional information, please see <a href='http://broadinstitute.github.io/picard/picard-metric-definitions.html#SequencingArtifactMetrics.PreAdapterDetailMetrics'>the PreAdapterDetailMetrics documentation</a>, the <a href='http://broadinstitute.github.io/picard/picard-metric-definitions.html#SequencingArtifactMetrics.PreAdapterSummaryMetrics'>the PreAdapterSummaryMetrics documentation</a>, the <a href='http://broadinstitute.github.io/picard/picard-metric-definitions.html#SequencingArtifactMetrics.BaitBiasDetailMetrics'>the BaitBiasDetailMetrics documentation</a>, and the <a href='http://broadinstitute.github.io/picard/picard-metric-definitions.html#SequencingArtifactMetrics.BaitBiasSummaryMetrics'>the BaitBiasSummaryMetrics documentation</a>. <hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>INTERVALS (File)</td><td>An optional list of intervals to restrict analysis to.  Default value: null. </td></tr>
<tr><td>DB_SNP (File)</td><td>VCF format dbSNP file, used to exclude regions around known polymorphisms from analysis.  Default value: null. </td></tr>
<tr><td>MINIMUM_QUALITY_SCORE (Integer)</td><td>The minimum base quality score for a base to be included in analysis.  Default value: 20. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>MINIMUM_MAPPING_QUALITY (Integer)</td><td>The minimum mapping quality score for a base to be included in analysis.  Default value: 30. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>MINIMUM_INSERT_SIZE (Integer)</td><td>The minimum insert size for a read to be included in analysis.  Default value: 60. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>MAXIMUM_INSERT_SIZE (Integer)</td><td>The maximum insert size for a read to be included in analysis. Set to 0 to have no maximum.  Default value: 600. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>INCLUDE_UNPAIRED (Boolean)</td><td>Include unpaired reads. If set to true then all paired reads will be included as well - MINIMUM_INSERT_SIZE and MAXIMUM_INSERT_SIZE will be ignored.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>TANDEM_READS (Boolean)</td><td>Set to true if mate pairs are being sequenced from the same strand, i.e. they're expected to face the same direction.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>USE_OQ (Boolean)</td><td>When available, use original quality scores for filtering.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>CONTEXT_SIZE (Integer)</td><td>The number of context bases to include on each side of the assayed base.  Default value: 1. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>CONTEXTS_TO_PRINT (String)</td><td>If specified, only print results for these contexts in the detail metrics output. However, the summary metrics output will still take all contexts into consideration.  Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>FILE_EXTENSION (String)</td><td>Append the given file extension to all metric file names (ex. OUTPUT.pre_adapter_summary_metrics.EXT). None if null  Default value: null. </td></tr>
<tr><td>INPUT (File)</td><td>Input SAM or BAM file.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>File to write the output to.  Required. </td></tr>
<tr><td>ASSUME_SORTED (Boolean)</td><td>If true (default), then the sort order in the header file will be ignored.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>STOP_AFTER (Long)</td><td>Stop after processing N reads, mainly for debugging.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
