<h3 id="GenotypeConcordance">GenotypeConcordance</h3>
<section>
<p>Evaluate genotype concordance between callsets.This tool evaluates the concordance between genotype calls for samples in different callsets where one is being considered as the truth (aka standard, or reference) and the other as the call that is being evaluated for accuracy. <br /><h4>Usage example:</h4><pre>java -jar picard.jar GenotypeConcordance \<br />      CALL_VCF=input.vcf \<br />      CALL_SAMPLE=sample_name \<br />      O=gc_concordance.vcf \<br />      TRUTH_VCF=truth_set.vcf \<br />      TRUTH_SAMPLE=truth_sample#</pre><h4>Output Metrics:</h4><ul><li>Output metrics include GenotypeConcordanceContingencyMetrics, GenotypeConcordanceSummaryMetrics, and GenotypeConcordanceDetailMetrics.  For each set of metrics, the data is broken into separate sections for SNPs and INDELs.  Note that only SNP and INDEL variants are considered, MNP, Symbolic, and Mixed classes of variants are not included. </li><li>GenotypeConcordanceContingencyMetrics enumerate the constituents of each contingent in a callset including true-positive (TP), true-negative (TN), false-positive (FP), and false-negative (FN) calls. See http://broadinstitute.github.io/picard/picard-metric-definitions.html#GenotypeConcordanceContingencyMetrics for more details.</li><li>GenotypeConcordanceDetailMetrics include the numbers of SNPs and INDELs for each contingent genotype as well as the number of validated genotypes. See http://broadinstitute.github.io/picard/picard-metric-definitions.html#GenotypeConcordanceDetailMetrics for more details.</li><li>GenotypeConcordanceSummaryMetrics provide specific details for the variant caller performance on a callset including: values for sensitivity, specificity, and positive predictive values. See http://broadinstitute.github.io/picard/picard-metric-definitions.html#GenotypeConcordanceSummaryMetrics for more details.</li></ul><br /><br />Useful definitions applicable to alleles and genotypes:<br /> <ul><li>Truthset - A callset (typically in VCF format) containing variant calls and genotypes that have been cross-validated with multiple technologies e.g. Genome In A Bottle Consortium (GIAB) (https://sites.stanford.edu/abms/giab)</li><li>TP - True positives are variant calls that match a 'truthset'</li><li>FP - False-positives are reference sites miscalled as variant</li><li>FN - False-negatives are variant sites miscalled as reference</li><li>TN - True negatives are correctly called reference sites</li><li>Validated genotypes - are TP sites where the exact genotype (HET or HOM-VAR) has been validated </li> </ul><hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>TRUTH_VCF (File)</td><td>The VCF containing the truth sample  Required. </td></tr>
<tr><td>CALL_VCF (File)</td><td>The VCF containing the call sample  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>Basename for the two metrics files that are to be written. Resulting files will be &lt;OUTPUT>.genotype_concordance_summary_metrics  and &lt;OUTPUT>.genotype_concordance_detail_metrics.  Required. </td></tr>
<tr><td>TRUTH_SAMPLE (String)</td><td>The name of the truth sample within the truth VCF  Required. </td></tr>
<tr><td>CALL_SAMPLE (String)</td><td>The name of the call sample within the call VCF  Required. </td></tr>
<tr><td>INTERVALS (File)</td><td>One or more interval list files that will be used to limit the genotype concordance.  Note - if intervals are specified, the VCF files must be indexed.  Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>INTERSECT_INTERVALS (Boolean)</td><td>If true, multiple interval lists will be intersected. If false multiple lists will be unioned.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>MIN_GQ (Integer)</td><td>Genotypes below this genotype quality will have genotypes classified as LowGq.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>MIN_DP (Integer)</td><td>Genotypes below this depth will have genotypes classified as LowDp.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>OUTPUT_ALL_ROWS (Boolean)</td><td>If true, output all rows in detailed statistics even when count == 0.  When false only output rows with non-zero counts.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>USE_VCF_INDEX (Boolean)</td><td>If true, use the VCF index, else iterate over the entire VCF.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>MISSING_SITES_HOM_REF (Boolean)</td><td>Default is false, which follows the GA4GH Scheme. If true, missing sites in the truth set will be treated as HOM_REF sites and sites missing in both the truth and call sets will be true negatives. Useful when hom ref sites are left out of the truth set. This flag can only be used with a high confidence interval list.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
</table>
</section>
