<h3 id="IntervalListTools">IntervalListTools</h3>
<section>
<p>Manipulates interval lists.  This tool offers multiple interval list file manipulation capabilities include sorting, merging, subtracting, padding, customizing, and other set-theoretic operations. If given one or more inputs, the default operation is to merge and sort them.  Other options e.g. interval subtraction are controlled by the arguments. The tool lists intervals with respect to a reference sequence.<br /><br />Both interval_list and VCF files are accepted as input. The interval_list file format is relatively simple and reflects the SAM alignment format to a degree.  A SAM style header must be present in the file that lists the sequence records against which the intervals are described.  After the header, the file then contains records, one per line in text format with the following values tab-separated: <pre>     -Sequence name (SN) <br />     -Start position (1-based)** <br />     -End position (1-based, end inclusive) <br />     -Strand (either + or -) <br />     -Interval name (ideally unique names for intervals)</pre>The coordinate system of interval_list files is such that the first base or position in a sequence is position "1".<h4>Usage example:</h4><pre>java -jar picard.jar IntervalListTools \<br />      I=input.interval_list \<br />      SI=input_2.interval_list \<br />      O=new.interval_list</pre><hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>INPUT (File)</td><td>One or more interval lists. If multiple interval lists are provided the output is theresult of merging the inputs. Supported formats are interval_list and VCF.  Default value: null. This option must be specified at least 1 times. </td></tr>
<tr><td>OUTPUT (File)</td><td>The output interval list file to write (if SCATTER_COUNT is 1) or the directory into which to write the scattered interval sub-directories (if SCATTER_COUNT > 1)  Default value: null. </td></tr>
<tr><td>PADDING (Integer)</td><td>The amount to pad each end of the intervals by before other operations are undertaken. Negative numbers are allowed and indicate intervals should be shrunk. Resulting intervals &lt; 0 bases long will be removed. Padding is applied to the interval lists &lt;b> before &lt;/b> the ACTION is performed.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>UNIQUE (Boolean)</td><td>If true, merge overlapping and adjacent intervals to create a list of unique intervals. Implies SORT=true  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>SORT (Boolean)</td><td>If true, sort the resulting interval list by coordinate.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>ACTION (Action)</td><td>Action to take on inputs.  Default value: CONCAT. This option can be set to 'null' to clear the default value. Possible values: {
<p>CONCAT (The concatenation of all the INPUTs, no sorting or merging of overlapping/abutting intervals implied. Will result in an unsorted list unless requested otherwise.)
<p>UNION (Like CONCATENATE but with UNIQUE and SORT implied, the result being the set-wise union of all INPUTS.)
<p>INTERSECT (The sorted, uniqued set of all loci that are contained in all of the INPUTs.)
<p>SUBTRACT (Subtracts SECOND_INPUT from INPUT. The resulting loci are there in INPUT that are not in SECOND_INPUT)
<p>SYMDIFF (Find loci that are in INPUT or SECOND_INPUT but are not in both.)
<p>} </td></tr>
<tr><td>SECOND_INPUT (File)</td><td>Second set of intervals for SUBTRACT and DIFFERENCE operations.  Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>COMMENT (String)</td><td>One or more lines of comment to add to the header of the output file.  Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>SCATTER_COUNT (Integer)</td><td>The number of files into which to scatter the resulting list by locus; in some situations, fewer intervals may be emitted.  Note - if > 1, the resultant scattered intervals will be sorted and uniqued.  The sort will be inverted if the INVERT flag is set.  Default value: 1. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>INCLUDE_FILTERED (Boolean)</td><td>Whether to include filtered variants in the vcf when generating an interval list from vcf  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>BREAK_BANDS_AT_MULTIPLES_OF (Integer)</td><td>If set to a positive value will create a new interval list with the original intervals broken up at integer multiples of this value.  Set to 0 to NOT break up intervals  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>SUBDIVISION_MODE (Mode)</td><td>Do not subdivide   Default value: INTERVAL_SUBDIVISION. This option can be set to 'null' to clear the default value. Possible values: {INTERVAL_SUBDIVISION, BALANCING_WITHOUT_INTERVAL_SUBDIVISION, BALANCING_WITHOUT_INTERVAL_SUBDIVISION_WITH_OVERFLOW} </td></tr>
<tr><td>INVERT (Boolean)</td><td>Produce the inverse list  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
</table>
</section>
