<h3 id="MarkDuplicates">MarkDuplicates</h3>
<section>
<p>Identifies duplicate reads.  This tool locates and tags duplicate reads (both PCR and optical/sequencing-driven) in a BAM or SAM file, where
duplicate reads are defined as originating from the same original fragment of DNA. Duplicates are identified as read
pairs having identical 5' positions (coordinate and strand) for both reads in a mate pair (and optinally, matching
unique molecular identifier reads; see BARCODE_TAG option). Optical, or more broadly Sequencing, duplicates are
duplicates that appear clustered together spatially during sequencing and can arise from optical/imagine-processing
artifacts or from bio-chemical processes during clonal amplification and sequencing; they are identified using the
READ_NAME_REGEX and the OPTICAL_DUPLICATE_PIXEL_DISTANCE options.

The tool's main output is a new SAM or BAM file in which duplicates have been identified in the SAM flags field, or
optionally removed (see REMOVE_DUPLICATE and REMOVE_SEQUENCING_DUPLICATES), and optionally marked with a duplicate type
in the 'DT' optional attribute. In addition, it also outputs a metrics file containing the numbers of
READ_PAIRS_EXAMINED, UNMAPPED_READS, UNPAIRED_READS, UNPAIRED_READ DUPLICATES, READ_PAIR_DUPLICATES, and
READ_PAIR_OPTICAL_DUPLICATES.
<br /><h4>Usage example:</h4><pre>java -jar picard.jar MarkDuplicates I=input.bam \<br />      O=marked_duplicates.bam M=marked_dup_metrics.txt<br /></pre><hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP (Integer)</td><td>This option is obsolete. ReadEnds will always be spilled to disk.  Default value: 50000. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>MAX_FILE_HANDLES_FOR_READ_ENDS_MAP (Integer)</td><td>Maximum number of file handles to keep open when spilling read ends to disk. Set this number a little lower than the per-process maximum number of file that may be open. This number can be found by executing the 'ulimit -n' command on a Unix system.  Default value: 8000. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>SORTING_COLLECTION_SIZE_RATIO (Double)</td><td>This number, plus the maximum RAM available to the JVM, determine the memory footprint used by some of the sorting collections.  If you are running out of memory, try reducing this number.  Default value: 0.25. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>BARCODE_TAG (String)</td><td>Barcode SAM tag (ex. BC for 10X Genomics)  Default value: null. </td></tr>
<tr><td>READ_ONE_BARCODE_TAG (String)</td><td>Read one barcode SAM tag (ex. BX for 10X Genomics)  Default value: null. </td></tr>
<tr><td>READ_TWO_BARCODE_TAG (String)</td><td>Read two barcode SAM tag (ex. BX for 10X Genomics)  Default value: null. </td></tr>
<tr><td>REMOVE_SEQUENCING_DUPLICATES (Boolean)</td><td>If true remove 'optical' duplicates and other duplicates that appear to have arisen from the sequencing process instead of the library preparation process, even if REMOVE_DUPLICATES is false. If REMOVE_DUPLICATES is true, all duplicates are removed and this option is ignored.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>TAGGING_POLICY (DuplicateTaggingPolicy)</td><td>Determines how duplicate types are recorded in the DT optional attribute.  Default value: DontTag. This option can be set to 'null' to clear the default value. Possible values: {DontTag, OpticalOnly, All} </td></tr>
<tr><td>INPUT (String)</td><td>One or more input SAM or BAM files to analyze. Must be coordinate sorted.  Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>OUTPUT (File)</td><td>The output file to write marked records to  Required. </td></tr>
<tr><td>METRICS_FILE (File)</td><td>File to write duplication metrics to  Required. </td></tr>
<tr><td>REMOVE_DUPLICATES (Boolean)</td><td>If true do not write duplicates to the output file instead of writing them with appropriate flags set.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>ASSUME_SORTED (Boolean)</td><td>If true, assume that the input file is coordinate sorted even if the header says otherwise.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>DUPLICATE_SCORING_STRATEGY (ScoringStrategy)</td><td>The scoring strategy for choosing the non-duplicate among candidates.  Default value: SUM_OF_BASE_QUALITIES. This option can be set to 'null' to clear the default value. Possible values: {SUM_OF_BASE_QUALITIES, TOTAL_MAPPED_REFERENCE_LENGTH} </td></tr>
<tr><td>PROGRAM_RECORD_ID (String)</td><td>The program record ID for the @PG record(s) created by this program. Set to null to disable PG record creation.  This string may have a suffix appended to avoid collision with other program record IDs.  Default value: MarkDuplicates. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>PROGRAM_GROUP_VERSION (String)</td><td>Value of VN tag of PG record to be created. If not specified, the version will be detected automatically.  Default value: null. </td></tr>
<tr><td>PROGRAM_GROUP_COMMAND_LINE (String)</td><td>Value of CL tag of PG record to be created. If not supplied the command line will be detected automatically.  Default value: null. </td></tr>
<tr><td>PROGRAM_GROUP_NAME (String)</td><td>Value of PN tag of PG record to be created.  Default value: MarkDuplicates. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>COMMENT (String)</td><td>Comment(s) to include in the output file's header.  Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>READ_NAME_REGEX (String)</td><td>Regular expression that can be used to parse read names in the incoming SAM file. Read names are parsed to extract three variables: tile/region, x coordinate and y coordinate. These values are used to estimate the rate of optical duplication in order to give a more accurate estimated library size. Set this option to null to disable optical duplicate detection. The regular expression should contain three capture groups for the three variables, in order. It must match the entire read name. Note that if the default regex is specified, a regex match is not actually done, but instead the read name  is split on colon character. For 5 element names, the 3rd, 4th and 5th elements are assumed to be tile, x and y values. For 7 element names (CASAVA 1.8), the 5th, 6th, and 7th elements are assumed to be tile, x and y values.  Default value: <optimized capture of last three ':' separated fields as numeric values>. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>OPTICAL_DUPLICATE_PIXEL_DISTANCE (Integer)</td><td>The maximum offset between two duplicate clusters in order to consider them optical duplicates. The default is appropriate for unpatterned versions of the Illumina platform. For the patterned flowcell models, 2500 is moreappropriate. For other platforms and models, users should experiment to find what works best.  Default value: 100. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
