<h3 id="MarkDuplicatesWithMateCigar">MarkDuplicatesWithMateCigar</h3>
<section>
<p>Identifies duplicate reads, accounting for mate CIGAR.  This tool locates and tags duplicate reads (both PCR and optical) in a BAM or SAM file, where duplicate reads are defined as originating from the same original fragment of DNA, taking into account the CIGAR string of read mates. <br /><br />It is intended as an improvement upon the original MarkDuplicates algorithm, from which it differs in several ways, includingdifferences in how it breaks ties. It may be the most effective duplicate marking program available, as it handles all cases including clipped and gapped alignments and locates duplicate molecules using mate cigar information. However, please note that it is not yet used in the Broad's production pipeline, so use it at your own risk. <br /><br />Note also that this tool will not work with alignments that have large gaps or deletions, such as those from RNA-seq data.  This is due to the need to buffer small genomic windows to ensure integrity of the duplicate marking, while large skips (ex. skipping introns) in the alignment records would force making that window very large, thus exhausting memory. <br /><h4>Usage example:</h4><pre>java -jar picard.jar MarkDuplicatesWithMateCigar \<br />      I=input.bam \<br />      O=mark_dups_w_mate_cig.bam \<br />      M=mark_dups_w_mate_cig_metrics.txt</pre><hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>MINIMUM_DISTANCE (Integer)</td><td>The minimum distance to buffer records to account for clipping on the 5' end of the records.Set this number to -1 to use twice the first read's read length (or 100, whichever is smaller).  Default value: -1. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>SKIP_PAIRS_WITH_NO_MATE_CIGAR (Boolean)</td><td>Skip record pairs with no mate cigar and include them in the output.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>BLOCK_SIZE (Integer)</td><td>The block size for use in the coordinate-sorted record buffer.  Default value: 100000. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>INPUT (String)</td><td>One or more input SAM or BAM files to analyze. Must be coordinate sorted.  Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>OUTPUT (File)</td><td>The output file to write marked records to  Required. </td></tr>
<tr><td>METRICS_FILE (File)</td><td>File to write duplication metrics to  Required. </td></tr>
<tr><td>REMOVE_DUPLICATES (Boolean)</td><td>If true do not write duplicates to the output file instead of writing them with appropriate flags set.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>ASSUME_SORTED (Boolean)</td><td>If true, assume that the input file is coordinate sorted even if the header says otherwise.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>DUPLICATE_SCORING_STRATEGY (ScoringStrategy)</td><td>The scoring strategy for choosing the non-duplicate among candidates.  Default value: TOTAL_MAPPED_REFERENCE_LENGTH. This option can be set to 'null' to clear the default value. Possible values: {SUM_OF_BASE_QUALITIES, TOTAL_MAPPED_REFERENCE_LENGTH} </td></tr>
<tr><td>PROGRAM_RECORD_ID (String)</td><td>The program record ID for the @PG record(s) created by this program. Set to null to disable PG record creation.  This string may have a suffix appended to avoid collision with other program record IDs.  Default value: MarkDuplicates. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>PROGRAM_GROUP_VERSION (String)</td><td>Value of VN tag of PG record to be created. If not specified, the version will be detected automatically.  Default value: null. </td></tr>
<tr><td>PROGRAM_GROUP_COMMAND_LINE (String)</td><td>Value of CL tag of PG record to be created. If not supplied the command line will be detected automatically.  Default value: null. </td></tr>
<tr><td>PROGRAM_GROUP_NAME (String)</td><td>Value of PN tag of PG record to be created.  Default value: MarkDuplicatesWithMateCigar. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>COMMENT (String)</td><td>Comment(s) to include in the output file's header.  Default value: null. This option may be specified 0 or more times. </td></tr>
<tr><td>READ_NAME_REGEX (String)</td><td>Regular expression that can be used to parse read names in the incoming SAM file. Read names are parsed to extract three variables: tile/region, x coordinate and y coordinate. These values are used to estimate the rate of optical duplication in order to give a more accurate estimated library size. Set this option to null to disable optical duplicate detection. The regular expression should contain three capture groups for the three variables, in order. It must match the entire read name. Note that if the default regex is specified, a regex match is not actually done, but instead the read name  is split on colon character. For 5 element names, the 3rd, 4th and 5th elements are assumed to be tile, x and y values. For 7 element names (CASAVA 1.8), the 5th, 6th, and 7th elements are assumed to be tile, x and y values.  Default value: <optimized capture of last three ':' separated fields as numeric values>. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>OPTICAL_DUPLICATE_PIXEL_DISTANCE (Integer)</td><td>The maximum offset between two duplicate clusters in order to consider them optical duplicates. The default is appropriate for unpatterned versions of the Illumina platform. For the patterned flowcell models, 2500 is moreappropriate. For other platforms and models, users should experiment to find what works best.  Default value: 100. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
