<h3 id="QualityScoreDistribution">QualityScoreDistribution</h3>
<section>
<p>Chart the distribution of quality scores.  This tool is used for determining the overall "quality" for a library in a given run. To that effect, it outputs a chart and tables indicating the range of quality scores and the total numbers of bases corresponding to those scores. Options include plotting the distribution of all of the reads, only the aligned reads, or reads that have passed the Illumina Chastity filter thresholds as described <a href='http://gatkforums.broadinstitute.org/discussion/6329/pf-reads-illumina-chastity-filter'>here</a>.<br /> <br /><h4>Note on base quality score options</h4>If the quality score of read bases has been modified in a previous data processing step such as <a href='http://gatkforums.broadinstitute.org/discussion/44/base-quality-score-recalibration-bqsr'>GATK Base Recalibration</a> and an OQ tag is available, this tool can be set to plot the OQ value as well as the primary quality value for the evaluation. <br /><h4>Usage example:</h4><pre>java -jar picard.jar QualityScoreDistribution \<br />      I=input.bam \<br />      O=qual_score_dist.txt \<br />      CHART=qual_score_dist.pdf</pre><hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>CHART_OUTPUT (File)</td><td>A file (with .pdf extension) to write the chart to.  Required. </td></tr>
<tr><td>ALIGNED_READS_ONLY (Boolean)</td><td>If set to true calculate mean quality over aligned reads only.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>PF_READS_ONLY (Boolean)</td><td>If set to true calculate mean quality over PF reads only.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INCLUDE_NO_CALLS (Boolean)</td><td>If set to true, include quality for no-call bases in the distribution.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INPUT (File)</td><td>Input SAM or BAM file.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>File to write the output to.  Required. </td></tr>
<tr><td>ASSUME_SORTED (Boolean)</td><td>If true (default), then the sort order in the header file will be ignored.  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>STOP_AFTER (Long)</td><td>Stop after processing N reads, mainly for debugging.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
</table>
</section>
