<a id="ReorderSam"/>
<h3>ReorderSam</h3>
<section>
<p>Not to be confused with SortSam which sorts a SAM or BAM file with a valid sequence dictionary, ReorderSam reorders reads in a SAM/BAM file to match the contig ordering in a provided reference file, as determined by exact name matching of contigs.  Reads mapped to contigs absent in the new reference are dropped. Runs substantially faster if the input is an indexed BAM file.</p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>INPUT (File)</td><td>Input file (bam or sam) to extract reads from.  Required. </td></tr>
<tr><td>OUTPUT (File)</td><td>Output file (bam or sam) to write extracted reads to.  Required. </td></tr>
<tr><td>REFERENCE (File)</td><td>Reference sequence to reorder reads to match.  A sequence dictionary corresponding to the reference fasta is required.  Create one with CreateSequenceDictionary.jar.  Required. </td></tr>
<tr><td>ALLOW_INCOMPLETE_DICT_CONCORDANCE (Boolean)</td><td>If true, then allows only a partial overlap of the BAM contigs with the new reference sequence contigs.  By default, this tool requires a corresponding contig in the new reference for each read contig  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>ALLOW_CONTIG_LENGTH_DISCORDANCE (Boolean)</td><td>If true, then permits mapping from a read contig to a new reference contig with the same name but a different length.  Highly dangerous, only use if you know what you are doing.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
</table>
</section>
