<h3 id="SamToFastq">SamToFastq</h3>
<section>
<p>Converts a SAM or BAM file to FASTQ.  This tool extracts read sequences and base quality scores from the input SAM/BAM file and outputs them in FASTQ format. This can be used (by way of a pipe) to run BWA MEM on unmapped BAM (uBAM) files.<br /><h4>Usage example:</h4><pre>java -jar picard.jar SamToFASTQ \<br />     I=input.bam \<br />     FASTQ=output.fastq</pre><hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>INPUT (File)</td><td>Input SAM/BAM file to extract reads from  Required. </td></tr>
<tr><td>FASTQ (File)</td><td>Output FASTQ file (single-end fastq or, if paired, first end of the pair FASTQ).  Required.  Cannot be used in conjuction with option(s) OUTPUT_PER_RG (OPRG)</td></tr>
<tr><td>SECOND_END_FASTQ (File)</td><td>Output FASTQ file (if paired, second end of the pair FASTQ).  Default value: null.  Cannot be used in conjuction with option(s) OUTPUT_PER_RG (OPRG)</td></tr>
<tr><td>UNPAIRED_FASTQ (File)</td><td>Output FASTQ file for unpaired reads; may only be provided in paired-FASTQ mode  Default value: null.  Cannot be used in conjuction with option(s) OUTPUT_PER_RG (OPRG)</td></tr>
<tr><td>OUTPUT_PER_RG (Boolean)</td><td>Output a FASTQ file per read group (two FASTQ files per read group if the group is paired).  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false}  Cannot be used in conjuction with option(s) SECOND_END_FASTQ (F2) UNPAIRED_FASTQ (FU) FASTQ (F)</td></tr>
<tr><td>RG_TAG (String)</td><td>The read group tag (PU or ID) to be used to output a FASTQ file per read group.  Default value: PU. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>OUTPUT_DIR (File)</td><td>Directory in which to output the FASTQ file(s).  Used only when OUTPUT_PER_RG is true.  Default value: null. </td></tr>
<tr><td>RE_REVERSE (Boolean)</td><td>Re-reverse bases and qualities of reads with negative strand flag set before writing them to FASTQ  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INTERLEAVE (Boolean)</td><td>Will generate an interleaved fastq if paired, each line will have /1 or /2 to describe which end it came from  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>INCLUDE_NON_PF_READS (Boolean)</td><td>Include non-PF reads from the SAM file into the output FASTQ files. PF means 'passes filtering'. Reads whose 'not passing quality controls' flag is set are non-PF reads. See GATK Dictionary for more info.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
<tr><td>CLIPPING_ATTRIBUTE (String)</td><td>The attribute that stores the position at which the SAM record should be clipped  Default value: null. </td></tr>
<tr><td>CLIPPING_ACTION (String)</td><td>The action that should be taken with clipped reads: 'X' means the reads and qualities should be trimmed at the clipped position; 'N' means the bases should be changed to Ns in the clipped region; and any integer means that the base qualities should be set to that value in the clipped region.  Default value: null. </td></tr>
<tr><td>READ1_TRIM (Integer)</td><td>The number of bases to trim from the beginning of read 1.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>READ1_MAX_BASES_TO_WRITE (Integer)</td><td>The maximum number of bases to write from read 1 after trimming. If there are fewer than this many bases left after trimming, all will be written.  If this value is null then all bases left after trimming will be written.  Default value: null. </td></tr>
<tr><td>READ2_TRIM (Integer)</td><td>The number of bases to trim from the beginning of read 2.  Default value: 0. This option can be set to 'null' to clear the default value. </td></tr>
<tr><td>READ2_MAX_BASES_TO_WRITE (Integer)</td><td>The maximum number of bases to write from read 2 after trimming. If there are fewer than this many bases left after trimming, all will be written.  If this value is null then all bases left after trimming will be written.  Default value: null. </td></tr>
<tr><td>QUALITY (Integer)</td><td>End-trim reads using the phred/bwa quality trimming algorithm and this quality.  Default value: null. </td></tr>
<tr><td>INCLUDE_NON_PRIMARY_ALIGNMENTS (Boolean)</td><td>If true, include non-primary alignments in the output.  Support of non-primary alignments in SamToFastq is not comprehensive, so there may be exceptions if this is set to true and there are paired reads with non-primary alignments.  Default value: false. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
</table>
</section>
