<h3 id="SplitVcfs">SplitVcfs</h3>
<section>
<p>Splits SNPs and INDELs into separate files.  This tool reads in a VCF or BCF file and writes out the SNPs and INDELs it contains to separate files. The headers of the two output files will be identical and index files will be created for both outputs. If records other than SNPs or INDELs are present, set the STRICT option to "false", otherwise the tool will raise an exception and quit. <br /><h4>Usage example:</h4><pre>java -jar picard.jar SplitVcfs \<br />      I=input.vcf \<br />      SNP_OUTPUT=snp.vcf \<br />      INDEL_OUTPUT=indel.vcf \<br />      STRICT=false</pre><hr /></p>
<table>
<tr><th>Option</th><th>Description</th></tr>
<tr><td>INPUT (File)</td><td>The VCF or BCF input file  Required. </td></tr>
<tr><td>SNP_OUTPUT (File)</td><td>The VCF or BCF file to which SNP records should be written. The file format is determined by file extension.  Required. </td></tr>
<tr><td>INDEL_OUTPUT (File)</td><td>The VCF or BCF file to which indel records should be written. The file format is determined by file extension.  Required. </td></tr>
<tr><td>SEQUENCE_DICTIONARY (File)</td><td>The index sequence dictionary to use instead of the sequence dictionaries in the input files  Default value: null. </td></tr>
<tr><td>STRICT (Boolean)</td><td>If true an exception will be thrown if an event type other than SNP or indel is encountered  Default value: true. This option can be set to 'null' to clear the default value. Possible values: {true, false} </td></tr>
</table>
</section>
